### PR TITLE
Tip 1171 created product

### PR DESCRIPTION
**Pull request by ahocquard**
**Description (for Contributor and Core Developer)**
For now, I kept the nullability of the events disable/enable as the identifier could not be set.

But maybe the corrcet solution is Mathias's one with a dedicated object setting the identifier.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php [hunk 1]
```diff
@@ -474,6 +478,11 @@ public function isEnabled()
      */
     public function setEnabled($enabled)
     {
+        if ($enabled === $this->enabled){
+            return;
+        }
+
+        $this->events[] = true === $enabled ? new EnabledProduct($this->identifier) : new DisabledProduct($this->identifier);
```
Comment (anaelChardan) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php]: More ProductCreated

Comment (anaelChardan) [reply src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php]: More ProductDisabled

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/Events/DisabledProduct.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+declare(strict_types=1);
+
+namespace Akeneo\Pim\Enrichment\Component\Product\Model\Events;
+
+/**
+ * @copyright 2019 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class DisabledProduct
```
Comment (anaelChardan) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/Events/DisabledProduct.php]: Reverse the order of all your subject + participe ;)

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php (hunk 1 above)
Comment (arnolanglade) [reply src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php]: What do you think about having single event `ProductStatusChanged` ?

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php [hunk 3]
```diff
@@ -765,6 +774,12 @@ public function isVariant(): bool
 
     public function popEvents(): array
     {
+        // as the id is generated by the database, there is no other way to detect it as there is no function called
+        // to create a product
+        if (null === $this->id) {
+            $this->events[] = new CreatedProduct($this->identifier);
```
Comment (arnolanglade) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php]: I think it should be `ProductCreated`

Comment (arnolanglade) [reply src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php]: I don't think this is the purpose of this PR but you might start to introduce value object like the product identifier. They could only be used by event at the beginning.

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php [hunk 4]
```diff
@@ -765,6 +774,12 @@ public function isVariant(): bool
 
     public function popEvents(): array
     {
+        // as the id is generated by the database, there is no other way to detect it as there is no function called
+        // to create a product
+        if (null === $this->id) {
```
Comment (arnolanglade) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php]: :sob:

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/Events/CreatedProduct.php [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+declare(strict_types=1);
+
+namespace Akeneo\Pim\Enrichment\Component\Product\Model\Events;
+
+/**
+ * @copyright 2019 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class CreatedProduct
```
Comment (arnolanglade) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/Events/CreatedProduct.php]: I think event classes shoud be final. You don't need to override them.

Comment (arnolanglade) [reply src/Akeneo/Pim/Enrichment/Component/Product/Model/Events/CreatedProduct.php]: Missing spec

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/Events/DisabledProduct.php (hunk 2 above)
Comment (arnolanglade) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/Events/DisabledProduct.php]: Missing spec

Comment (arnolanglade) [reply src/Akeneo/Pim/Enrichment/Component/Product/Model/Events/DisabledProduct.php]: I think this class should be named `ProductDisabled`

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/Events/EnabledProduct.php [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+declare(strict_types=1);
+
+namespace Akeneo\Pim\Enrichment\Component\Product\Model\Events;
+
+/**
+ * @copyright 2019 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class EnabledProduct
```
Comment (arnolanglade) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/Events/EnabledProduct.php]: I think this class should be named `ProductEnabled`

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/Events/CreatedProduct.php (hunk 5 above)
Comment (ahocquard) [reply src/Akeneo/Pim/Enrichment/Component/Product/Model/Events/CreatedProduct.php]: I don't spec anymore such dumb classes. And I don't care about syntax and privacy in a POC

Comment (arnolanglade) [reply src/Akeneo/Pim/Enrichment/Component/Product/Model/Events/CreatedProduct.php]: It is your choice, if you don't use phpspec for what it made for: design you code. You should consider to use phpunit

Comment (ahocquard) [reply src/Akeneo/Pim/Enrichment/Component/Product/Model/Events/CreatedProduct.php]: I choosed to use my IDE for such classes ;)

**Pull request merged**